### PR TITLE
feat(core): add workflow execute subcommand

### LINKED ISSUE #2130: add `renku workflow execute` command
Add a `renku workflow execute` command that supports having multiple providers, and add a CWL provider for start (reusing our current code as much as possible or consider using `toil` as an alternative to the current cwl reference implementation)

### PR DESCRIPTION
**Pull request by vigsterkr**
# Description

Fixes #2130

this PR only includes the following features:
  - add subcommand `workflow execute` based on the specs
  - use pluggy to implement a plugin based workflow provider/executor framework
  - create a default `cwltool` based workflow provider
  - fix cwl exporter to use parameter's `actual_value` property instead of `default_value` this way user-defined parameter overrides will be reflected in the exported cwl file

### CODE REVIEW

File: renku/core/management/workflow/value_resolution.py [hunk 1]
```diff
@@ -17,16 +17,18 @@
...
-from renku.core import errors
 from renku.core.models.workflow.composite_plan import CompositePlan
 from renku.core.models.workflow.parameter import ParameterMapping
-from renku.core.models.workflow.plan import Plan
+from renku.core.models.workflow.plan import AbstractPlan, Plan
 
 
-def apply_run_values(workflow: Union[CompositePlan, Plan], values: Dict[str, Any] = None) -> None:
-    """Applies values and default_values to a potentially nested workflow.
+class RunValues:
```
Comment (vigsterkr) [inline renku/core/management/workflow/value_resolution.py:28]: @Panaetius is this somewhat in line what you had in mind?

File: renku/core/commands/workflow.py [hunk 2]
```diff
@@ -274,30 +284,79 @@ def _export_workflow(
     if output:
         output = Path(output)
 
+    if values:
+        values = _safe_read_yaml(values)
+        rv = RunValues(workflow, values)
+        workflow = rv.apply()
+        if rv.missing_parameters:
+            communication.warn(
```
Comment (vigsterkr) [inline renku/core/commands/workflow.py:292]: i've tested this on a local instance and for some reason `communication.warn` doesn't do anything, i.e. i'm not getting getting anything on the stdout/stderr

Comment (Panaetius) [reply renku/core/commands/workflow.py:292]: Did you set up the communication provider in the command builder? Like https://github.com/SwissDataScienceCenter/renku-python/blob/master/renku/cli/dataset.py#L467 and https://github.com/SwissDataScienceCenter/renku-python/blob/master/renku/cli/dataset.py#L472

File: renku/core/management/workflow/value_resolution.py [hunk 3]
```diff
@@ -36,70 +38,80 @@ def apply_run_values(workflow: Union[CompositePlan, Plan], values: Dict[str, Any
...
+            if "parameters" in self._values:
+                # NOTE: Set mapping parameter values
+                self._apply_parameters_values()
 
-                apply_run_values(child_workflow, step)
+            if "steps" in self._values:
+                for name, step in self._values["steps"].items():
+                    child_workflow = next((w for w in self._plan.plans if w.name == name), None)
+                    if not child_workflow:
+                        self.missing_parameters.add(f"steps.{name}")
```
Comment (Panaetius) [inline renku/core/management/workflow/value_resolution.py:79]: Maybe we should differentiate between a child workflow not existing and a parameter on a child workflow not existing

Comment (Panaetius) [inline renku/core/management/workflow/value_resolution.py:79]:
```suggestion
                        self.missing_parameters.add(f"steps.{name}")
                        continue
```

File: renku/core/management/workflow/value_resolution.py [hunk 4]
```diff
@@ -36,70 +38,80 @@ def apply_run_values(workflow: Union[CompositePlan, Plan], values: Dict[str, Any
...
-        mapping = next((m for m in workflow.mappings if m.name == k), None)
+    def _apply_parameters_values(self) -> None:
+        """Apply values to mappings of a CompositePlan."""
+        for k, v in self._values["parameters"].items():
+            mapping = next((m for m in self._plan.mappings if m.name == k), None)
 
-        if not mapping:
-            raise errors.ParameterNotFoundError(k, workflow.name)
+            if not mapping:
+                self.missing_parameters.add(k)
```
Comment (Panaetius) [inline renku/core/management/workflow/value_resolution.py:112]:
```suggestion
                self.missing_parameters.add(k)
                continue
```

File: renku/core/management/workflow/value_resolution.py [hunk 5]
```diff
@@ -36,70 +38,80 @@ def apply_run_values(workflow: Union[CompositePlan, Plan], values: Dict[str, Any
...
+                    child_workflow = next((w for w in self._plan.plans if w.name == name), None)
+                    if not child_workflow:
+                        self.missing_parameters.add(f"steps.{name}")
 
-    # apply defaults
-    for mapping in workflow.mappings:
-        apply_parameter_defaults(mapping)
+                    rv = RunValues(child_workflow, step)
+                    _ = rv.apply()
+                    self.missing_parameters.update(rv.missing_parameters)
```
Comment (Panaetius) [inline renku/core/management/workflow/value_resolution.py:83]: I think we should prepend `step.name` to each entry here before updating, so `missing_parameters` contains `step.mystep.learning_rate`, not just `learning_rate`

**Review by Panaetius — CHANGES_REQUESTED** [review]

File: renku/core/commands/workflow.py (hunk 2 above)
Comment (vigsterkr) [reply renku/core/commands/workflow.py:292]: 🤦 this is the result of a quickfix. thnx!

**Comment by vigsterkr** [comment]
@Panaetius ready for (re)review. thanks!

**Comment by Panaetius** [comment]
You should enjoy your vacation :wink: 

Don't worry, we can take care of this PR while you're gone. And thanks for the changes

File: renku/core/commands/update.py [hunk 6]
```diff
@@ -143,7 +146,9 @@ def execute_workflow(
 
     started_at_time = local_now()
 
-    modified_outputs = _execute_workflow_helper(plans=plans, client=client)
+    # NOTE: Create a ``CompositePlan`` because ``workflow_covert`` expects it
+    workflow = CompositePlan(id=CompositePlan.generate_id(), plans=plans, name=f"plan-collection-{uuid.uuid4().hex}")
```
Comment (vigsterkr) [inline renku/core/commands/update.py:150]: i dont like this as there's a `CompositePlan` -> `List[Plan]` -> `CompositePlan` cycle going on here. but for the sake of having the execute feature in master i'd merge this and the toil PR in the meanwhile i'll finish the refactor of handling list of plans in the workflow execute/converter plugins

Comment (Panaetius) [reply renku/core/commands/update.py:150]: Me and Mohammad discussed this in the past, and it is unfortunate but a bit of a necessary feature to allow the same executor to be used with `workflow execute`, `rerun` and `update`. In the rerun and update case, there is only a list of plans as the source, not a CompositePlan, and we didn't want to modify `execute_workflow` as you're still working on that part.

But we should consider not building a CompositePlan in the `execute_workflow` method as it's not really needed.

File: renku/core/management/workflow/value_resolution.py [hunk 7]
```diff
@@ -17,89 +17,135 @@
...
-        if "steps" in values:
-            for name, step in values["steps"].items():
-                child_workflow = next((w for w in workflow.plans if w.name == name), None)
-                if not child_workflow:
-                    raise errors.ChildWorkflowNotFound(name, workflow.name)
+            if "steps" in self._values:
+                for name, step in self._values["steps"].items():
+                    child_workflow = next((w for w in self._plan.plans if w.name == name), None)
+                    if not child_workflow:
+                        raise errors.ChildWorkflowNotFound(name, self._plan.name)
```
Comment (mohammad-sdsc) [inline renku/core/management/workflow/value_resolution.py:110]: `ChildWorkflowNotFound` -> `ChildWorkflowNotFoundError`

File: renku/core/commands/workflow.py [hunk 8]
```diff
@@ -20,10 +20,11 @@
 
 from datetime import datetime
 from pathlib import Path
-from typing import List, Optional
+from typing import Any, Dict, List, Optional
 
 from renku.core import errors
 from renku.core.commands.format.workflow import WORKFLOW_FORMATS
+from renku.core.commands.update import execute_workflow
```
Comment (mohammad-sdsc) [inline renku/core/commands/workflow.py:27]: Should we move `execute_workflow` function to `renku/core/commands/workflow.py`?

File: renku/cli/workflow.py [hunk 9]
```diff
@@ -599,3 +607,58 @@ def outputs(ctx, paths):
...
+    type=click.Choice(available_workflow_providers(), case_sensitive=False),
+    help="The workflow engine to use.",
+)
+@click.option("config", "-c", "--config", metavar="<config file>", help="YAML file containing config for the provider.")
+@click.option(
+    "set_params",
+    "-s",
+    "--set",
+    multiple=True,
+    metavar="<parameter>=<value>",
```
Comment (mohammad-sdsc) [inline renku/cli/workflow.py:628]: How does this work with `CompositePlanValueResolver`? In there we look for a `parameters` or a `steps` keys. It's also probably good to add a test which uses this flag.

**Review by mohammad-sdsc — COMMENTED** [review]
Thanks Viktor! This looks very good. I've made a few comments; feel free to address them in a coming refactoring PR if you are doing one atm.

**Review by mohammad-sdsc — APPROVED** [review]
Thank you!

**Review by Panaetius — APPROVED** [review]

**Pull request merged**